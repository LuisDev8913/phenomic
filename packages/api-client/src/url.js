function postfix(url) {
  return `${url}.json`;
}

const protect = encodeURIComponent;

function url(config: Object): string {
  const root = config.root || "";
  if (typeof config === "string") {
    return [root, config].join("/");
  }
  if (typeof config.id === "string") {
    return postfix([root, config.path, "item", config.id].join("/"));
  }
  return postfix(
    [
      root,
      protect(config.path),
      `by-${protect(config.by)}`,
      protect(config.value || "1"),
      protect(config.order || "desc"),
      ...(config.limit ? [`limit-${protect(config.limit)}`] : []),
      ...(config.limit && config.after
        ? [`after-${protect(config.after)}`]
        : [])
    ].join("/")
  );
}

export default url;
