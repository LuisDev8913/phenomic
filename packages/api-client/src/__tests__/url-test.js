import url from "../url";

it("should build url for single item", () => {
  expect(
    url({
      path: "posts",
      id: "test"
    })
  ).toMatchSnapshot();
});

it("should build url for items by date", () => {
  expect(
    url({
      path: "posts",
      by: "date"
    })
  ).toMatchSnapshot();
});

it("should build url for specific tag", () => {
  expect(
    url({
      path: "posts",
      by: "tag",
      value: "test"
    })
  ).toMatchSnapshot();

  expect(
    url({
      path: "posts",
      by: "tag",
      value: "test",
      order: "asc"
    })
  ).toMatchSnapshot();

  expect(
    url({
      path: "posts",
      by: "tag",
      value: "test",
      order: "asc",
      limit: 10
    })
  ).toMatchSnapshot();

  expect(
    url({
      path: "posts",
      by: "tag",
      value: "test",
      order: "asc",
      limit: 10,
      after: "BASE64"
    })
  ).toMatchSnapshot();

  expect(
    url({
      root: "local",
      path: "posts",
      by: "tag",
      value: "test",
      order: "asc",
      limit: 10,
      after: "BASE64"
    })
  ).toMatchSnapshot();
});
