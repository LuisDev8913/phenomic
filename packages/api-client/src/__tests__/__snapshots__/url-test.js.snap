// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build url for items by date 1`] = `"/posts/by-date/1/desc.json"`;

exports[`should build url for single item 1`] = `"/posts/item/test.json"`;

exports[`should build url for specific tag 1`] = `"/posts/by-tag/test/desc.json"`;

exports[`should build url for specific tag 2`] = `"/posts/by-tag/test/asc.json"`;

exports[`should build url for specific tag 3`] = `"/posts/by-tag/test/asc/limit-10.json"`;

exports[`should build url for specific tag 4`] = `"/posts/by-tag/test/asc/limit-10/after-BASE64.json"`;

exports[`should build url for specific tag 5`] = `"local/posts/by-tag/test/asc/limit-10/after-BASE64.json"`;
